File: backend/src/chat/chat.controller.ts
import { Controller, Post, Get, Body, Param, Query, HttpException, HttpStatus } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiQuery, ApiBody } from '@nestjs/swagger';
import { ChatService } from './chat.service';
import { SendMessageDto } from './dto/send-message.dto';
import { UsersService } from '../users/users.service';

@ApiTags('chat')
@Controller('api/chat')
export class ChatController {
  constructor(
    private readonly chatService: ChatService,
    private readonly usersService: UsersService,
  ) {}

  @Post('send')
  @ApiOperation({ summary: 'Send a chat message' })
  @ApiResponse({ status: 201, description: 'Message sent successfully' })
  @ApiResponse({ status: 400, description: 'Invalid input data' })
  @ApiBody({ type: SendMessageDto })
  async sendMessage(@Body() dto: SendMessageDto) {
    try {
      console.log('💬 Chat send request received:', { 
        requestId: dto.requestId, 
        senderId: dto.senderId, 
        messageLength: dto.message?.length,
        dtoKeys: Object.keys(dto),
        dtoValues: Object.values(dto),
      });
      
      // Validate required fields
      if (!dto.senderId) {
        console.error('❌ senderId is missing from request body');
        throw new HttpException('senderId is required', HttpStatus.BAD_REQUEST);
      }
      if (!dto.requestId) {
        console.error('❌ requestId is missing from request body');
        throw new HttpException('requestId is required', HttpStatus.BAD_REQUEST);
      }
      if (!dto.message) {
        console.error('❌ message is missing from request body');
        throw new HttpException('message is required', HttpStatus.BAD_REQUEST);
      }
      
      // Get sender info
      const sender = await this.usersService.getUserById(dto.senderId);
      if (!sender) {
        console.error(`❌ User not found: ${dto.senderId}`);
        throw new HttpException(`User not found: ${dto.senderId}`, HttpStatus.NOT_FOUND);
      }

      console.log('✅ Sender found:', { id: sender.id, name: sender.name, role: sender.role });

      const message = await this.chatService.sendMessage(
        dto.requestId,
        dto.senderId,
        sender.name,
        sender.role,
        dto.message,
      );

      return {
        ...message,
        message: 'Message sent successfully',
      };
    } catch (error: any) {
      console.error('❌ Error in sendMessage:', error);
      if (error instanceof HttpException) {
        throw error;
      }
      throw new HttpException(
        error.message || 'Failed to send message',
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('messages/:requestId')
  @ApiOperation({ summary: 'Get all messages for a request' })
  @ApiParam({ name: 'requestId', description: 'Request ID' })
  @ApiResponse({ status: 200, description: 'List of messages' })
  @ApiResponse({ status: 404, description: 'Request not found' })
  async getMessages(
    @Param('requestId') requestId: string,
    @Query('userId') userId: string,
  ) {
    try {
      if (!userId) {
        throw new HttpException('userId query parameter is required', HttpStatus.BAD_REQUEST);
      }

      const messages = await this.chatService.getMessages(requestId, userId);
      return {
        messages,
        count: messages.length,
      };
    } catch (error: any) {
      console.error('❌ Error in getMessages:', error);
      if (error instanceof HttpException) {
        throw error;
      }
      throw new HttpException(
        error.message || 'Failed to get messages',
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('participants/:requestId')
  async getParticipants(@Param('requestId') requestId: string) {
    try {
      const participants = await this.chatService.getChatParticipants(requestId);
      if (!participants) {
        throw new HttpException(`Request not found: ${requestId}`, HttpStatus.NOT_FOUND);
      }

      // Get user names
      const needy = await this.usersService.getUserById(participants.needyId);
      const provider = participants.providerId
        ? await this.usersService.getUserById(participants.providerId)
        : null;

      return {
        needyId: participants.needyId,
        needyName: needy?.name || 'Unknown',
        providerId: participants.providerId,
        providerName: provider?.name || null,
      };
    } catch (error: any) {
      console.error('❌ Error in getParticipants:', error);
      if (error instanceof HttpException) {
        throw error;
      }
      throw new HttpException(
        error.message || 'Failed to get participants',
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('unread-counts/:userId')
  async getUnreadCounts(@Param('userId') userId: string) {
    try {
      const unreadCounts = await this.chatService.getUnreadCountsForUser(userId);
      const counts: Record<string, number> = {};
      unreadCounts.forEach((count, requestId) => {
        counts[requestId] = count;
      });
      return {
        success: true,
        unreadCounts: counts,
        totalUnread: Array.from(unreadCounts.values()).reduce((sum, count) => sum + count, 0),
      };
    } catch (error: any) {
      console.error('❌ Error in getUnreadCounts:', error);
      throw new HttpException(
        error.message || 'Failed to get unread counts',
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('mark-read/:requestId')
  async markAsRead(
    @Param('requestId') requestId: string,
    @Body('userId') userId: string,
  ) {
    try {
      if (!userId) {
        throw new HttpException('userId is required', HttpStatus.BAD_REQUEST);
      }
      this.chatService.markAsRead(requestId, userId);
      return {
        success: true,
        message: 'Messages marked as read',
      };
    } catch (error: any) {
      console.error('❌ Error in markAsRead:', error);
      if (error instanceof HttpException) {
        throw error;
      }
      throw new HttpException(
        error.message || 'Failed to mark as read',
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }
}

File: backend-dotnet/Controllers/ChatController.cs
using Microsoft.AspNetCore.Mvc;
using FuelMateBackend.Services;
using FuelMateBackend.DTOs;

namespace FuelMateBackend.Controllers;

[ApiController]
[Route("api/[controller]")]
[Tags("chat")]
public class ChatController : ControllerBase
{
    private readonly ChatService _chatService;
    private readonly UsersService _usersService;

    public ChatController(ChatService chatService, UsersService usersService)
    {
        _chatService = chatService;
        _usersService = usersService;
    }

    /// <summary>
    /// Send a chat message
    /// </summary>
    [HttpPost("send")]
    public async Task<IActionResult> SendMessage([FromBody] SendMessageDto dto)
    {
        try
        {
            Console.WriteLine($"💬 Chat send request received: requestId={dto.RequestId}, senderId={dto.SenderId}, messageLength={dto.Message?.Length}");

            if (string.IsNullOrEmpty(dto.SenderId))
            {
                Console.WriteLine("❌ senderId is missing from request body");
                return BadRequest("senderId is required");
            }

            if (string.IsNullOrEmpty(dto.RequestId))
            {
                Console.WriteLine("❌ requestId is missing from request body");
                return BadRequest("requestId is required");
            }

            if (string.IsNullOrEmpty(dto.Message))
            {
                Console.WriteLine("❌ message is missing from request body");
                return BadRequest("message is required");
            }

            var sender = await _usersService.GetUserById(dto.SenderId);
            if (sender == null)
            {
                Console.WriteLine($"❌ User not found: {dto.SenderId}");
                return NotFound($"User not found: {dto.SenderId}");
            }

            Console.WriteLine($"✅ Sender found: id={sender.Id}, name={sender.Name}, role={sender.Role}");

            var message = await _chatService.SendMessage(
                dto.RequestId,
                dto.SenderId,
                sender.Name,
                sender.Role,
                dto.Message);

            return Ok(new
            {
                message.Id,
                message.RequestId,
                message.SenderId,
                message.SenderName,
                message.SenderRole,
                message.Message,
                message.CreatedAt,
                successMessage = "Message sent successfully"
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error in sendMessage: {ex.Message}");
            return StatusCode(500, ex.Message);
        }
    }

    /// <summary>
    /// Get all messages for a request
    /// </summary>
    [HttpGet("messages/{requestId}")]
    public async Task<IActionResult> GetMessages(string requestId, [FromQuery] string? userId)
    {
        try
        {
            if (string.IsNullOrEmpty(userId))
            {
                return BadRequest("userId query parameter is required");
            }

            var messages = await _chatService.GetMessages(requestId, userId);

            return Ok(new
            {
                messages,
                count = messages.Count
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error in getMessages: {ex.Message}");
            return StatusCode(500, ex.Message);
        }
    }

    /// <summary>
    /// Get chat participants
    /// </summary>
    [HttpGet("participants/{requestId}")]
    public async Task<IActionResult> GetParticipants(string requestId)
    {
        try
        {
            var participants = await _chatService.GetChatParticipants(requestId);
            if (participants == null)
            {
                return NotFound($"Request not found: {requestId}");
            }

            var needy = await _usersService.GetUserById(participants.Value.needyId);
            var provider = participants.Value.providerId != null
                ? await _usersService.GetUserById(participants.Value.providerId)
                : null;

            return Ok(new
            {
                needyId = participants.Value.needyId,
                needyName = needy?.Name ?? "Unknown",
                providerId = participants.Value.providerId,
                providerName = provider?.Name
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error in getParticipants: {ex.Message}");
            return StatusCode(500, ex.Message);
        }
    }

    /// <summary>
    /// Get unread message counts for a user
    /// </summary>
    [HttpGet("unread-counts/{userId}")]
    public IActionResult GetUnreadCounts(string userId)
    {
        try
        {
            var unreadCounts = _chatService.GetUnreadCountsForUser(userId);
            var totalUnread = unreadCounts.Values.Sum();

            return Ok(new
            {
                success = true,
                unreadCounts,
                totalUnread
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error in getUnreadCounts: {ex.Message}");
            return StatusCode(500, ex.Message);
        }
    }

    /// <summary>
    /// Mark messages as read
    /// </summary>
    [HttpPost("mark-read/{requestId}")]
    public IActionResult MarkAsRead(string requestId, [FromBody] MarkReadDto dto)
    {
        try
        {
            if (string.IsNullOrEmpty(dto.UserId))
            {
                return BadRequest("userId is required");
            }

            _chatService.MarkAsRead(requestId, dto.UserId);

            return Ok(new
            {
                success = true,
                message = "Messages marked as read"
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error in markAsRead: {ex.Message}");
            return StatusCode(500, ex.Message);
        }
    }
}

